go-controller/pkg/ovn/libovsdbops/acl.go

```diff
@@ -0,0 +1,192 @@
+package libovsdbops
+
+import (
+	"fmt"
+	"reflect"
+
+	libovsdbclient "github.com/ovn-org/libovsdb/client"
+	libovsdb "github.com/ovn-org/libovsdb/ovsdb"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/nbdb"
+)
+
+// getACLName returns the ACL name if it has one otherwise returns
+// an empty string.
+func getACLName(acl *nbdb.ACL) string {
+	if acl.Name != nil {
+		return *acl.Name
+	}
+	return ""
+}
+
+// isEquivalentACL if it has same uuid, or if it has same name
+// and external ids, or if it has same priority, direction, match
+// and action.
+func isEquivalentACL(existing *nbdb.ACL, searched *nbdb.ACL) bool {
+	if searched.UUID != "" && existing.UUID == searched.UUID {
+		return true
+	}
+
+	eName := getACLName(existing)
+	sName := getACLName(searched)
+	// TODO if we want to support adding/removing external ids,
+	// we need to compare them differently, perhaps just the common subset
+	if eName != "" && eName == sName && reflect.DeepEqual(existing.ExternalIDs, searched.ExternalIDs) {
+		return true
+	}
+
+	return existing.Priority == searched.Priority &&
+		existing.Direction == searched.Direction &&
+		existing.Match == searched.Match &&
+		existing.Action == searched.Action
+}
+
+// findACL looks up the client cache for equivalent ACLs
+func findACL(nbClient libovsdbclient.Client, acl *nbdb.ACL) (string, error) {
+	if acl.UUID != "" && !IsNamedUUID(acl.UUID) {
+		return acl.UUID, nil
+	}
+
+	acls := []nbdb.ACL{}
+	err := nbClient.WhereCache(func(item *nbdb.ACL) bool {
+		return isEquivalentACL(item, acl)
+	}).List(&acls)
+	if err != nil {
+		return "", fmt.Errorf("can't find ACL by equivalence %+v: %v", *acl, err)
+	}
+
+	if len(acls) > 1 {
+		return "", fmt.Errorf("unexpectedly found multiple equivalent ACLs: %+v", acls)
+	}
+
+	if len(acls) == 1 {
+		return acls[0].UUID, nil
+	}
+
+	return "", nil
+}
+
+func BuildACL(name, direction nbdb.ACLDirection, priority int, match string, action nbdb.ACLAction, meter string, severity nbdb.ACLSeverity, log bool, externalIds map[string]string) *nbdb.ACL {
```

Comment:
Genuine question: does this function need to exist? I find it super hard to read. Would it be better for CreateOrUpdateACLs to do the name / uuid munging directly?

Then you could just directly instantiate structs in the calling functions.

Reply:
Also, `name` should have type `string`.

Reply:
One more thing - the counter will, of course, reset every time ovn-kube restarts. What is the risk of collision?

Reply:
Yes, we could move the uuid thing to `CreateOrUpdate`. This still comes handy as a place to common handling on the ACLs (or equally for other types) data like the name crop or anything else that comes up in the future. I think this sets a pattern to have single instance of these types of functions vs having them scattered  around the code which we already had and which might end up appearing again.

Reply:
To prevent nil-able parameters we could have different build functions.

Reply:
`BuildACL` handles named UUIDs so the can be used to reference the ACL from other places like PGs that will be included in the same transaction.

Reply:
For the sake of clarity could you add `name string, ...`, it's currently a string because it's "shadowed" from `nbdb.ACLDirection` which is a string, but it's not clear when reading the function signature